rfd/0147-repo-fork-syncing.md

```diff
@@ -0,0 +1,60 @@
+---
+authors: Mike Jensen (mike.jensen@goteleport.com)
+state: draft
+---
+ 
+# RFD 144 - Automatic Repo Syncing
+ 
+## What
+ 
+This RFD describes how we can sync a repo continuously and in an automatic way.  This can be used to provide private copies of a repo.
+ 
+## Why
+ 
+We currently face a few frictions which could be addressed through a more robust repo sync mechanism:
+1. Our internal `teleport-private` fork used for security development must be manually updated
+2. Some tooling (Dependabot) does not allow scanning for non-default branches, necessitating manual efforts to manage our release branches
+3. Because much of the CI and scanning configuration is committed into the repo, we face noise from duplicate alerts when we use a repo for testing or as an internal fork
+ 
+## Details
+ 
+As such we need a solution that can provide the following:
+* Automatic and regular updates to a repo to ensure it's in sync
+* The ability to apply changes to the upstream repo (for example disable Actions that we don't want running on the repo copy, or enable additional scanning / actions)
+ 
+### Branch Structure
+ 
+* `main` - This will be the repo's default branch.  This branch will contain the upstream history in addition to commits needed for the sync process or other custom changes on the repo copy
+* `sync/upstream-[UPSTREAM_BRANCH_NAME]` - This branch will be an exact copy of the upstream branch.  If there is a need to create a PR in this repo the history should be based off this rather than `main`
+* `sync/rebase` - This branch is where our custom changes will be committed.  It will be rebased on to the upstream changes before being committed into `main`
+ 
+#### Branch Protections
+ 
+The only branch that will be protected is the `sync/rebase` branch.  Other branches are expected to allow force pushes so they can be forcefully synced based off the upstream state.
+ 
+### High Level Implementation
+ 
+At a high level there are two primary components to this implementation:
+1. A GitHub action which will be scheduled as well as triggered on changes to the `sync/rebase` branch.  This action will perform the actual rebase action and commit the results to the respective branches above.
+2. A script that is invoked when the rebase can not be applied cleanly.  This can allow some custom rebase logic as a means to reduce how often the `sync/rebase` branch needs to be manually updated to avoid rebase conflicts.  A simple example is that workflow automations not desired in the repo copy may be removed in the `sync/rebase` branch.  That means that any changes to those committed workflows would result in a conflict.  However it can be safely resolved that any modified files removed in the `sync/rebase` branch can be resolved by simply removing.
```

Comment:
What happens when this script fails?  Who gets notified (if anyone) and how?

Reply:
I get an email when CodeQL fails in `teleport` currently, or when report generation happens in `security-reports`.  I assume everyone who watches the repo is getting these emails on failure?

That was my plan here as well, let me know if you have a better recommendation